api/src/main/java/org/apache/iceberg/expressions/BoundAggregate.java

```diff
@@ -44,4 +66,100 @@ public Type type() {
       return term().type();
     }
   }
+
+  public String columnName() {
+    if (op() == Operation.COUNT_STAR) {
+      return "*";
+    } else {
+      return ref().name();
+    }
+  }
+
+  public String describe() {
+    switch (op()) {
+      case COUNT_STAR:
+        return "count(*)";
+      case COUNT:
+        return "count(" + ExpressionUtil.describe(term()) + ")";
+      case MAX:
+        return "max(" + ExpressionUtil.describe(term()) + ")";
+      case MIN:
+        return "min(" + ExpressionUtil.describe(term()) + ")";
+      default:
+        throw new UnsupportedOperationException("Unsupported aggregate type: " + op());
+    }
+  }
+
+  <V> V safeGet(Map<Integer, V> map, int key) {
+    return safeGet(map, key, null);
+  }
+
+  <V> V safeGet(Map<Integer, V> map, int key, V defaultValue) {
+    if (map != null) {
+      return map.getOrDefault(key, defaultValue);
+    }
+
+    return null;
+  }
+
+  interface Aggregator<R> {
+    void update(StructLike struct);
+
+    void update(DataFile file);
+
+    R result();
+
+    boolean hasValue();
+  }
+
+  abstract static class NullSafeAggregator<T, R> implements Aggregator<R> {
+    private final BoundAggregate<T, R> aggregate;
+    private boolean hasValue = true;
+
+    NullSafeAggregator(BoundAggregate<T, R> aggregate) {
+      this.aggregate = aggregate;
+    }
+
+    protected abstract void update(R value);
+
+    protected abstract R current();
+
+    @Override
+    public void update(StructLike struct) {
+      if (hasValue) {
+        R value = aggregate.eval(struct);
+        if (aggregate.canPushDown) {
+          update(value);
+        } else {
+          this.hasValue = false;
+        }
+      }
+    }
+
+    @Override
+    public void update(DataFile file) {
+      if (hasValue) {
+        R value = aggregate.eval(file);
```

Comment:
Rather than setting `canPushDown` in the aggregator, I think that this needs to distinguish between a `null` value and a missing value. To do that, there are a few options, but I think the cleanest is to add a `hasValue(DataFile file)` method to check before calling `eval`. If that returns false, then `hasValue` is set to `false` and no more aggregation is done.

In addition, this method should only call `update` when values are non-null. Aggregation skips null values for count, min, and max, and count-star will return the number of rows in a file and is always non-null.